src/gc/gc.cpp

```diff
@@ -12327,33 +12332,44 @@ BOOL gc_heap::a_fit_segment_end_p (int gen_number,
     }
 #endif //BACKGROUND_GC
 
-    uint8_t* old_alloc;
-    old_alloc = allocated;
 #ifdef FEATURE_LOH_COMPACTION
     if (gen_number == (max_generation + 1))
     {
-        make_unused_array (old_alloc, loh_pad);
-        old_alloc += loh_pad;
+        make_unused_array (allocated, loh_pad);
         allocated += loh_pad;
         limit -= loh_pad;
     }
 #endif //FEATURE_LOH_COMPACTION
 
-#if defined (VERIFY_HEAP) && defined (_DEBUG)
-        ((void**) allocated)[-1] = 0;     //clear the sync block
-#endif //VERIFY_HEAP && _DEBUG
-    allocated += limit;
-
     dprintf (3, ("found fit at end of seg: %Ix", old_alloc));
 
+    uint8_t* old_alloc;
+    old_alloc = allocated;
+
 #ifdef BACKGROUND_GC
     if (cookie != -1)
     {
+        allocated += limit;
         bgc_loh_alloc_clr (old_alloc, limit, acontext, flags, align_const, cookie, TRUE, seg);
     }
     else
 #endif //BACKGROUND_GC
-    {
+    {      
+        // In a contiguous AC case with GC_ALLOC_ZEROING_OPTIONAL, deduct unspent space from the limit to clear only what is necessary.
+        if (flags & GC_ALLOC_ZEROING_OPTIONAL &&
+            (allocated == acontext->alloc_limit || allocated == acontext->alloc_limit + Align (min_obj_size, align_const)))
+        {
+            limit -= (allocated - acontext->alloc_ptr);
```

Comment:
>limit -= (allocated - acontext->alloc_ptr); [](start = 12, length = 43)

why reduce limit here? we already knew that we could fit the limit we asked for, might as well just give it out anyway. I don't understand why this needs to be done because in adjust_limit_clr we will naturally only advance acontext->alloc_ptr by a min_obj_size when alloc_limit was end of seg, ie, we do not make large free objects (we just made the pad).

Reply:
if we give more than necessary, we will have to zero the space beyond the start + size. That would be roughly the size of the leftover in the current AC. In 4K allocation the leftover could be up to 4K. Clearing that much extra is noticeable.
That would be ok for a regular allocation, but for GC_ALLOC_ZEROING_OPTIONAL we are trying to not clear more than needed.

---
In reply to: [287178268](https://github.com/dotnet/coreclr/pull/24504#discussion_r287178268) [](ancestors = 287178268)

Reply:
I reduce the limit so we would only have to clear the pad at the end. (since that is not consumed by the object, we have to clean it). 
If I do not reduce, we will have to clear additional "allocated - acontext->alloc_ptr" bytes since that would not be consumed by the object and whatever is left in the AC needs to be clean as the next allocation may just use that space without clearing.

---
In reply to: [287180883](https://github.com/dotnet/coreclr/pull/24504#discussion_r287180883) [](ancestors = 287180883,287178268)

Reply:
ahh, I thought you said we were wasting space in the AC when we talked about this. I misunderstood then. this seems like an optimization targeted at a very narrow scenario almost not worth the complexity. I think I'm ok with keeping this if you feel strongly about it but I could definitely go the other way.

also you don't need to check for gen_number in the if block because for LOH allocations the if statement will not be true.

Reply:
I was not sure about gen_number. Thanks!

Yeah, I'd like to keep this. According to VTune these extra expenses are noticeable. 
That is how I found this - it was not obvious from the code, but profiles indicated that we are spending time in memset even when doing Uninitialized allocations.

I also tried to combine this with the similar logic inside alloc_clr. That promised to make it simpler, but at the end did not work because we must update "allocated" here while we still holding the heap lock.

---
In reply to: [287183348](https://github.com/dotnet/coreclr/pull/24504#discussion_r287183348) [](ancestors = 287183348)

Reply:
Will have to remove unnecessary gen_number check.

---
In reply to: [287189803](https://github.com/dotnet/coreclr/pull/24504#discussion_r287189803) [](ancestors = 287189803,287183348)